threedworld-mit/tdw
# Added parameter mesh_scale to set_flex_cloth_actor

# How to test


1. Make sure your version of TDW is up to date (`pip3 install tdw -U` if installed from PyPi or `git pull` if installed from this repo)
2. Download, extract, and run [this build](https://www.dropbox.com/s/odo6bbjc67r6rbq/TDW_Linux_cloth_scale.tar.gz?dl=0)
3. Run this controller:

```python
from tdw.controller import Controller
from tdw.tdw_utils import TDWUtils

c = Controller(launch_build=False)
solid_id = c.get_unique_id()
cloth_id = c.get_unique_id()
c.communicate([{"$type": "load_scene",
                 "scene_name": "ProcGenScene"},
                TDWUtils.create_empty_room(12, 12),
                {"$type": "create_flex_container",
                 "collision_distance": 0.001,
                 "static_friction": 0.1,
                 "dynamic_friction": 0.1,
                 "iteration_count": 5,
                 "substep_count": 8,
                 "radius": 0.225,
                 "damping": 0,
                 "drag": 0},
               {"$type": "create_avatar",
                "type": "A_Img_Caps_Kinematic",
                "id": "a"},
               {"$type": "teleport_avatar_to",
                "position": {"x": 3, "y": 1.5, "z": -2}},
               {"$type": "look_at_position",
                "position": {"x": 0, "y": 0.25, "z": 0}},
               c.get_add_object(model_name="cube",
                                library="models_flex.json",
                                object_id=solid_id),
               {"$type": "set_flex_solid_actor",
                "id": solid_id,
                "mass_scale": 100.0,
                "particle_spacing": 0.1},
               {"$type": "assign_flex_container",
                "id": solid_id,
                "container_id": 0},
               c.get_add_object(model_name="cloth_square",
                                library="models_special.json",
                                object_id=cloth_id,
                                position={"x": 0, "y": 2, "z": 0}),
               {"$type": "set_flex_cloth_actor",
                "id": cloth_id,
                "mass_scale": 1,
                "mesh_tesselation": 1,
                "tether_stiffness": 0.5,
                "bend_stiffness": 0.2,
                "self_collide": False,
                "stretch_stiffness": 0.1,
                "pressure": 0.1,
                "mesh_scale": {"x": 0.5, "y": 2, "z": 1.3}},
               {"$type": "assign_flex_container",
                "id": cloth_id,
                "container_id": 0}])
for i in range(200):
    c.communicate([])
```

# Changes

### Command API

#### Modified Commands

| Command                | Modification                                                 |
| ---------------------- | ------------------------------------------------------------ |
| `set_flex_cloth_actor` | Added optional parameter `mesh_scale` to set the scale of the object's mesh. Default value: (1, 1, 1). This fixes a bug in which it was impossible to scale a Flex cloth object. |

### Documentation

#### Modified Documentation

| Document  | Modification                                                 |
| --------- | ------------------------------------------------------------ |
| `flex.md` | Updated example lists of commands.<br>Added a section for scaling objects.<br>Added a note about collision detection. |
